File: example/helloworld.c
/*
 * Copyright (C) 2017 Jianhui Zhao <jianhuizhao329@gmail.com>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301
 * USA
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <uhttpd.h>
#include <libubox/ulog.h>

static void hello_action(struct uh_client *cl)
{
    int body_len = 0;

    cl->send_header(cl, 200, "OK", -1);
    cl->append_header(cl, "Myheader", "Hello");
    cl->header_end(cl);

    cl->chunk_printf(cl, "<h1>Hello Libuhttpd %s</h1>", UHTTPD_VERSION_STRING);
    cl->chunk_printf(cl, "<h1>REMOTE_ADDR: %s</h1>", cl->get_peer_addr(cl));
    cl->chunk_printf(cl, "<h1>URL: %s</h1>", cl->get_url(cl));
    cl->chunk_printf(cl, "<h1>PATH: %s</h1>", cl->get_path(cl));
    cl->chunk_printf(cl, "<h1>QUERY: %s</h1>", cl->get_query(cl));
    cl->chunk_printf(cl, "<h1>VAR name: %s</h1>", cl->get_var(cl, "name"));
    cl->chunk_printf(cl, "<h1>BODY:%s</h1>", cl->get_body(cl, &body_len));
    cl->request_done(cl);
}

static void usage(const char *prog)
{
    fprintf(stderr, "Usage: %s [option]\n"
        "          -p port  # Default port is 8080\n"
        "          -s       # SSl on\n"
        "          -v       # verbose\n", prog);
    exit(1);
}

int main(int argc, char **argv)
{
    struct uh_server *srv = NULL;
    bool verbose = false;
    bool ssl = false;
    int port = 8080;
    int opt;

    while ((opt = getopt(argc, argv, "p:vs")) != -1) {
        switch (opt)
        {
        case 'p':
            port = atoi(optarg);
            break;
        case 's':
            ssl = true;
            break;
        case 'v':
            verbose = true;
            break;
        default: /* '?' */
            usage(argv[0]);
        }
    }

    if (!verbose)
        ulog_threshold(LOG_ERR);
    
    ULOG_INFO("libuhttpd version: %s\n", UHTTPD_VERSION_STRING);

    uloop_init();

    srv = uh_server_new("0.0.0.0", port);
    if (!srv)
        goto done;

#if (!UHTTPD_SSL_SUPPORT)
    if (ssl)
        ULOG_ERR("SSl is not compiled in\n");
#else
    if (ssl && srv->ssl_init(srv, "server-key.pem", "server-cert.pem") < 0)
        goto done;
#endif

    ULOG_INFO("Listen on: %s *:%d\n", srv->ssl ? "https" : "http", port);

    srv->add_action(srv, "/hello", hello_action);
    
    uloop_run();
done:
    uloop_done();
    srv->free(srv);
    
    return 0;
}

File: example/lua/helloworld.lua
#!/usr/bin/env lua

--[[
  Copyright (C) 2017 Jianhui Zhao <jianhuizhao329@gmail.com>
 
  This program is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.
 
  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.
 
  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301
  USA
 --]]

local uloop = require "uloop"
local uh = require "uhttpd"

local verbose = true
local port = 8914

uloop.init()

-- LOG_DEBUG LOG_INFO LOG_ERR
if not verbose then
    uh.set_log_threshold(uh.LOG_ERR)
end

uh.log(uh.LOG_INFO, "uhttpd version:" .. uh.VERSION)

local srv = uh.new(port)

-- srv:ssl_init("uhttpd.crt", "uhttpd.key")
-- srv:set_options({docroot = "/home/zjh/www", index = "lua.html"})

uh.log(uh.LOG_INFO, "Listen on:" .. port)



local http_methods = {
    [uh.HTTP_METHOD_GET] = "GET",
    [uh.HTTP_METHOD_POST] = "POST",
    [uh.HTTP_METHOD_HEAD] = "HEAD"
}
local http_version = {
    [uh.HTTP_VER_09] = "HTTP/0.9",
    [uh.HTTP_VER_10] = "HTTP/1.0",
    [uh.HTTP_VER_11] = "HTTP/1.1"
}

srv:on_error404(function(cl, path)
    cl:send_header(200, "OK", -1)
    cl:header_end()
    cl:chunk_send(string.format("<h1>Libuhttpd-Lua: '%s' Not found</h1>", path))
    cl:request_done()
end)


srv:on_request(function(cl, path)
    if path ~= "/hello" then
        return uh.REQUEST_CONTINUE
    end

    cl:send_header(200, "OK", -1)
    cl:append_header("Myheader", "Hello")
    cl:header_end()

    cl:chunk_send(string.format("<h1>Hello Libuhttpd %s</h1>", uh.VERSION))
    cl:chunk_send(string.format("<h1>REMOTE_ADDR: %s</h1>", cl:get_remote_addr()))
    cl:chunk_send(string.format("<h1>METHOD: %s</h1>", http_methods[cl:get_http_method()]))
    cl:chunk_send(string.format("<h1>HTTP Version: %s</h1>", http_version[cl:get_http_version()]))
    cl:chunk_send(string.format("<h1>URL: %s</h1>", cl:get_url()))
    cl:chunk_send(string.format("<h1>QUERY: %s</h1>", cl:get_query() or ""))
    cl:chunk_send(string.format("<h1>Body: %s</h1>", cl:get_body() or ""))

    -- Get a http var
    local var_x = cl:get_var("x")
    cl:chunk_send(string.format("<h1>Var x: %s</h1>", var_x or ""))

    -- Get a http header
    local user_agent = cl:get_header("user-agent")
    cl:chunk_send(string.format("<h1>User-Agent: %s</h1>", user_agent))

    cl:chunk_send("<hr />")
    -- Get all http vars
    local vars = cl:get_var()
    for k, v in pairs(vars) do
        cl:chunk_send(string.format("<h1>%s: %s</h1>", k, v))
    end

    cl:chunk_send("<hr />")
    -- Get all http headers
    local headers = cl:get_header()
    for k, v in pairs(headers) do
        cl:chunk_send(string.format("<h1>%s: %s</h1>", k, v))
    end

    cl:request_done()

    return uh.REQUEST_DONE
end)


uloop.run()
